pkg/agent/controller/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,658 @@
+		}
+		if pc.Spec.Packet.TransportHeader.TCP.DstPort != nil {
+			packet.DestinationPort = uint16(*pc.Spec.Packet.TransportHeader.TCP.DstPort)
+		}
+	} else if pc.Spec.Packet.TransportHeader.UDP != nil {
+		if pc.Spec.Packet.TransportHeader.UDP.SrcPort != nil {
+			packet.SourcePort = uint16(*pc.Spec.Packet.TransportHeader.UDP.SrcPort)
+		}
+		if pc.Spec.Packet.TransportHeader.UDP.DstPort != nil {
+			packet.DestinationPort = uint16(*pc.Spec.Packet.TransportHeader.UDP.DstPort)
+		}
+	}
+	return packet, nil
+}
+
+func (c *Controller) syncPacketCapture(pcName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing PacketCapture", "name", pcName, "startTime", time.Since(startTime))
+	}()
+
+	pc, err := c.packetCaptureLister.Get(pcName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.cleanupPacketCapture(pcName)
+			return nil
+		}
+		return err
+	}
+
+	switch pc.Status.Phase {
+	case "":
+		err = c.initPacketCapture(pc)
+	case crdv1alpha1.PacketCaptureRunning:
+		err = c.checkPacketCaptureStatus(pc)
+	default:
+		c.cleanupPacketCapture(pcName)
+	}
+	return err
+
+}
+
+func (c *Controller) getUploaderByProtocol(protocol StorageProtocolType) (ftp.Uploader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+func (c *Controller) generatePacketsPathForServer(name string) string {
+	return name + ".pcapng"
+}
+
+func getDefaultFileServerAuth() *crdv1alpha1.BundleServerAuthConfiguration {
+	return &crdv1alpha1.BundleServerAuthConfiguration{
+		AuthType: crdv1alpha1.BasicAuthentication,
+		AuthSecret: &v1.SecretReference{
+			Name:      fileServerAuthSecretName,
+			Namespace: fileServerAuthSecretNamespace,
+		},
+	}
+}
+
+func (c *Controller) uploadPackets(pc *crdv1alpha1.PacketCapture, outputFile afero.File) error {
+	if pc.Spec.FileServer == nil {
+		klog.V(2).Info("No fileserver info found in PacketCapture, skip upload packets file")
+		return nil
+	}
+	klog.V(2).InfoS("Uploading captured packets for PacketCapture", "name", pc.Name)
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return fmt.Errorf("failed to upload support bundle while getting uploader: %v", err)
+	}
+	authConfig := getDefaultFileServerAuth()
+	serverAuth, err := ftp.ParseBundleAuth(*authConfig, c.kubeClient)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get authentication for the fileServer", "name", pc.Name, "authentication", authConfig)
+		return err
+	}
+	cfg := ftp.GenSSHClientConfig(serverAuth.BasicAuthentication.Username, serverAuth.BasicAuthentication.Password)
+	return uploader.Upload(pc.Spec.FileServer.URL, c.generatePacketsPathForServer(pc.Name), cfg, outputFile)
+}
+
+// initPacketCapture mark the PacketCapture as running
+func (c *Controller) initPacketCapture(pc *crdv1alpha1.PacketCapture) error {
+	err := c.updatePacketCaptureStatus(pc, crdv1alpha1.PacketCaptureRunning, "", 0)
```

Comment:
Should we mark it as Running directly?
I feel this should be a validatePacketCapture() instead of initPacketCapture() to set it as running directly.
validatePacketCapture can set the status as failure and update the validation failure in the reason.
The failure during package capture should be updated during the PacketCaptureRunning status.
@tnqn  any suggestion?

Reply:
The suggestion makes sense to me.

Reply:
Most of the `validation` has been handled by the CRD schema. Maybe the `getPodIP` part can be checked first?